addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java

```diff
@@ -0,0 +1,1299 @@
+     *
+     * @param callbackUrl
+     * @param reqCode
+     * @return
+     */
+    public SpotifyWebAPIAuthResult authenticate(String callbackUrl, String reqCode) {
+
+        final String authString = Base64.getEncoder()
+                .encodeToString(String.format("%s:%s", clientId, clientSecret).getBytes());
+
+        logger.debug("Sending Spotify Web API autorization request");
+
+        String content = String.format("grant_type=authorization_code&code=%s&redirect_uri=%s", reqCode, callbackUrl);
+
+        String contentType = "application/x-www-form-urlencoded";
+
+        for (int i = 0; i < HTTP_CLIENT_RETRY_COUNT; i++) {
+            logger.debug("Spotify Web API call attempt: {}", i + 1);
+            try {
+                ContentResponse response = httpClient.POST("https://accounts.spotify.com/api/token")
+                        .header("Authorization", "Basic " + authString)
+                        .content(new StringContentProvider(content), contentType)
+                        .timeout(HTTP_CLIENT_TIMEOUT, TimeUnit.SECONDS).send();
+
+                logger.debug("Response Code: {}", response.getStatus());
+                logger.trace("Response Data: {}", response.getContentAsString());
+
+                if (response.getStatus() == 200) {
+                    Gson gson = new Gson();
+                    SpotifyWebAPIAuthResult test = gson.fromJson(response.getContentAsString(),
+                            SpotifyWebAPIAuthResult.class);
+                    accessToken = test.accessToken;
+                    tokenValidity = test.getExpiresIn();
+                    return test;
+                } else if (response.getStatus() == 400) {
+                    logger.error("Response: {} - verify that Spotify Client ID and Client Secret are correct!",
+                            response.getContentAsString());
+
+                }
+            } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                logger.warn("Error calling Spotify Web API for authorization - no accessToken!", e);
+            }
+            logger.debug("Attempt {} failed.", i + 1);
+        }
+        logger.error("Giving up on accessing Spotify Web API. Check network connectivity!");
+        return null;
+    }
+
+    /**
+     * Schedule timely refreshes of the accessToken.
+     *
+     * @return false if refresh is not successful
+     */
+    public boolean scheduleAccessTokenRefresh() {
+
+        if (future != null && !future.isCancelled()) {
+            // stop previous refresh thread
+            future.cancel(true);
+        }
+
+        try {
+            refreshToken();
+        } catch (java.lang.NullPointerException npe) {
+            // ignore
+        }
+
+        if (accessToken == null) {
+            return false;
+        }
+
+        // TODO: Find a more suitable to retrieve token validity if it changes after being scheduled? Scheduling refresh
```

Comment:
There is a TODO, what should be done and can it be completed before merging the PR?